plugins/ccgx/fu-ccgx-dmc-devx-device.c

```diff
@@ -0,0 +1,275 @@
+		FuDevice *proxy = fu_device_get_proxy(device);
+		if (!fu_strtoull(value, &tmp, 0, G_MAXUINT32, error))
+			return FALSE;
+		if (fu_device_get_version_raw(proxy) != tmp) {
+			g_debug("overriding composite version from %u to %u from %s",
+				(guint)fu_device_get_version_raw(proxy),
+				(guint)tmp,
+				fu_device_get_id(device));
+			fu_device_set_version_from_uint32(proxy, tmp);
+		}
+		return TRUE;
+	}
+	g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "no supported");
+	return FALSE;
+}
+
+static const gchar *
+fu_ccgx_dmc_devx_device_type_to_name(DmcDevxDeviceType device_type)
+{
+	if (device_type == DMC_DEVX_DEVICE_TYPE_CCG3)
+		return "CCG3";
+	if (device_type == DMC_DEVX_DEVICE_TYPE_DMC)
+		return "DMC";
+	if (device_type == DMC_DEVX_DEVICE_TYPE_CCG4)
+		return "CCG4";
+	if (device_type == DMC_DEVX_DEVICE_TYPE_CCG5)
+		return "CCG5";
+	if (device_type == DMC_DEVX_DEVICE_TYPE_HX3)
+		return "HX3";
+	return "Unknown";
+}
+
+static gboolean
+fu_ccgx_dmc_devx_device_probe(FuDevice *device, GError **error)
+{
+	FuCcgxDmcDevxDevice *self = FU_CCGX_DMC_DEVX_DEVICE(device);
+	FuDevice *proxy = fu_device_get_proxy(device);
+	DmcDevxStatus *status = &self->status;
+	DmcDevxDeviceType device_version_type = fu_ccgx_dmc_devx_device_version_type(self);
+	gsize offset = 0;
+	g_autofree gchar *logical_id = g_strdup_printf("0x%02x", status->component_id);
+	g_autofree gchar *version = NULL;
+
+	fu_device_set_name(device, fu_ccgx_dmc_devx_device_type_to_name(status->device_type));
+	fu_device_set_logical_id(device, logical_id);
+
+	/* for the version number */
+	if (status->current_image == 0x01)
+		offset = 4;
+	else if (status->current_image == 0x02)
+		offset = 8;
+
+	/* version, if possible */
+	if (device_version_type == DMC_DEVX_DEVICE_TYPE_DMC) {
+		version = fu_ccgx_dmc_devx_status_version_dmc_bfw(status, offset);
+		fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_QUAD);
+	} else if (device_version_type == DMC_DEVX_DEVICE_TYPE_HX3) {
+		version = fu_ccgx_dmc_devx_status_version_hx3(status, offset);
+		fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_TRIPLET);
+		fu_device_set_version(device, version);
+	}
+	if (version != NULL) {
+		fu_device_set_version(device, version);
+		fu_device_add_instance_strsafe(device, "VER", version);
+	}
+
+	/* add GUIDs */
+	fu_device_add_instance_strup(device,
+				     "TYPE",
+				     fu_ccgx_dmc_devx_device_type_to_string(status->device_type));
+	fu_device_add_instance_u8(device, "CID", status->component_id);
+	fu_device_add_instance_u16(device, "VID", fu_usb_device_get_vid(FU_USB_DEVICE(proxy)));
+	fu_device_add_instance_u16(device, "PID", fu_usb_device_get_pid(FU_USB_DEVICE(proxy)));
+	fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "CID", NULL);
+	fu_device_build_instance_id_quirk(device, NULL, "USB", "VID", "PID", "CID", "TYPE", NULL);
+	fu_device_build_instance_id_quirk(device, NULL, "USB", "VID", "PID", "CID", "VER", NULL);
```

Comment:
This should only be if VER isn't null right?

Reply:
`fu_device_build_instance_id_quirk()` just skips the instance ID if any one of the keys is unset.